### DISCUSSION TITLE
iperf3 generating traffic slightly lower on one machine

### DISCUSSION THREAD

**Discussion by 77yu77** [esnet/iperf#1762] [General]
Hi,the scenario requires me to use iperf3 to generate traffic. And on the same machine (such as between containers), I wrote a program to send traffic to compare with iperf3. When one CPU is fully occupied, the average bandwidth is 28.35Gbits/s, while the iperf3 test shows 25.7Gbit/s with the same conditions. Does this mean that iperf3 may have some extra CPU resource consumption that causes the bandwidth to be lower?

**Comment by bmah888** [comment]
What version of iperf3 and what command-line arguments (both client and server)?

Was this result repeatable over multiple runs?

**Reply by 77yu77** [reply]
iperf 3.9 (cJSON 1.7.13).
I think it is repeatable.
Should I use the latest version to have a try?

**Reply by 77yu77** [reply]
I just use the iperf3.17 to run for 100 seconds. The avg is 25.7Gbits/s too. The client CPU cos is 100%-102%.

**Reply by davidBar-On** [reply]
Are you testing using TCP or UDP?  Can you list the client and server commands you are using (i.e. the `iperf3 -s/-c ...` commands)?
Also, in the program you wrote, what is the message size you are using?  I.e. how many byte you are sending in each `write()`/`send()` or whatever function you are using to send each message?

**Reply by 77yu77** [reply]
About the iperf3, I use TCP. And just use the normal iperf3 commands to send message. And in my program I use the same way as iperf3, that is the readentropy function to get data from /dev/urandom. The buffer size I use is 1024 * 1024.

**Reply by 77yu77** [reply]
I did not set a limit on the traffic limit of iperf3, but to test its maximum performance.

**Reply by davidBar-On** [reply]
> The buffer size I use is 1024 * 1024

Do you mean that you send 1MB messages, i.e. set iperf3 option `-l 1024K`? 

If yes, you should run iperf3 using `-l 1024K`. 

If not, are you using the default iperf3 message length?  Can you run the client (or server) using `-V` and send the "byte blocks" value in the "Starting Test: .... nnnnn byte blocks, ....." line?  This is the length of the `write()` message.  In addition what is the size of the `write()` (or `send()`, etc.) you are using in your program?

In any case, to compare the tests properly, **the iperf3 `-l` value and the size of the message written in your program should be the same**.

**Reply by 77yu77** [reply]
Oh, I use the  `-l 1024K` to set the TCP buffer the same as my program, it get the same result as mine. I forgot it. Previously, I thought that the bandwidth obtained by using iperf without any restrictions was the maximum bandwidth, but it seems that this idea was wrong.

**Reply by 77yu77** [reply]
Thanks you very much.

**Reply by davidBar-On** [reply]
😄
The reason that the `-l` impact the performance is actually as you suggested in the PR description: "Does this mean that iperf3 may have some extra CPU resource consumption that causes the bandwidth to be lower?".  iperf3 has a CPU overhead before each `write()` and each call to write has its overhead, so if the message length is smaller more writes are performed.

**Reply by 77yu77** [reply]
Yes, so when running locally, the data is read and written in memory, so the transmission speed is faster, resulting in this part of the overhead being very obvious. When transmitting across machines, the bottleneck is the network card and link, so the TCP buffer has little effect (when the network card and link are normal)? I tried my best to modify the code but could not reach 10Gbits/s for 10 Gbps network card.

**Reply by davidBar-On** [reply]
> When transmitting across machines, the bottleneck is the network card and link, so the TCP buffer has little effect (when the network card and link are normal)?

The actual machine to machine permanence depends on many different parameters, including the ones you mentioned, so usually it is not possible to understand the behavior without analyzing the traffic between the different components.  Different buffers sizes, Switches between the machines, etc. also impact the performance.

If your NIC supports any [TCP offload engine](https://en.wikipedia.org/wiki/TCP_offload_engine#Large_send_offload) methods (e.g. GRO) that may help.  With iperf3 you can also try the Zero-copy option (`-Z`) to see if that method helps.

**Reply by 77yu77** [reply]
Thanks! I will have a try.
